mllib/src/main/scala/org/apache/spark/mllib/feature/SignificantSelector.scala

```diff
@@ -0,0 +1,127 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.feature
+
+import scala.collection.mutable
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.mllib.linalg.{DenseVector, SparseVector, Vector, Vectors}
+import org.apache.spark.rdd.RDD
+
+/**
+ * :: Experimental ::
+ * Model to extract significant indices from vector.
+ *
+ * Significant indices is vector's index that has different value for different vectors.
+ *
+ * For example, when you use HashingTF they create big sparse vector,
+ * and this code convert to smallest vector that don't include same values indices for all vectors.
+ *
+ * @param indices array of significant indices.
+ */
+@Experimental
+class SignificantSelectorModel(val indices: Array[Int]) extends VectorTransformer {
+
+  /**
+   * Applies transformation on a vector.
+   *
+   * @param vector vector to be transformed.
+   * @return transformed vector.
+   */
+  override def transform(vector: Vector): Vector = vector match {
+    case DenseVector(vs) =>
+      Vectors.dense(indices.map(vs))
+
+    case SparseVector(s, ids, vs) =>
+      var sv_idx = 0
+      var new_idx = 0
+      val elements = new mutable.ListBuffer[(Int, Double)]()
+      
+      for (idx <- indices) {
+        while (sv_idx < ids.length && ids(sv_idx) < idx) {
+          sv_idx += 1
+        }
+        if (sv_idx < ids.length && ids(sv_idx) == idx) {
+          elements += ((new_idx, vs(sv_idx)))
+          sv_idx += 1
+        }
+        new_idx += 1
+      }
+      
+      Vectors.sparse(indices.length, elements)
+
+    case v =>
+      throw new IllegalArgumentException("Don't support vector type " + v.getClass)
+  }
+}
+
+/**
+ * :: Experimental ::
+ * Specialized model for equivalent vectors
+ */
+@Experimental
+class SignificantSelectorEmptyModel extends SignificantSelectorModel(Array[Int]()) {
+  
+  val empty_vector = Vectors.dense(Array[Double]())
+  
+  override def transform(vector: Vector): Vector = empty_vector
+}
+
+/**
+ * :: Experimental ::
+ * Create Significant selector.
+ */
+@Experimental
+class SignificantSelector() {
+
+  /**
+   * Returns a significant vector indices selector.
+   *
+   * @param sources an `RDD[Vector]` containing the vectors.
+   */
+  def fit(sources: RDD[Vector]): SignificantSelectorModel = {
+    val sources_count = sources.count()
+    val significant_indices = sources.flatMap {
+        case DenseVector(vs) =>
+          vs.zipWithIndex
+        case SparseVector(_, ids, vs) =>
+          vs.zip(ids)
+        case v =>
+          throw new IllegalArgumentException("Don't support vector type " + v.getClass)
+      }
+      .map(e => (e.swap, 1))
+      .reduceByKey(_ + _)
+      .map { case ((idx, value), count) => (idx, (value, count))}
+      .groupByKey()
+      .mapValues { e =>
+        val values = e.groupBy(_._1)
+        val sum = e.map(_._2).sum
+
+        values.size + (if (sum == sources_count || values.contains(0.0)) 0 else 1)
```

Comment:
Oh, this is hack for case when you has RDD what include dense and sparse vector.

Sparse vector hasn't got zero elements (`0d`) and `values.size` for sparse vector has count only for different non zero value.

If you has in RDD sparse and dense vector and last one has zero element where sparse vectro hasn't got element significant understand it's different values but isn't in fact.

For example, let's see following code:
```scala
    val vectors = sc.parallelize(List(
      Vectors.dense(0.0, 2.0, 3.0, 4.0),
      Vectors.dense(0.0, 2.0, 3.0, 4.0),
      Vectors.dense(0.0, 2.0, 3.0, 4.0),
      Vectors.sparse(4, Seq((1, 3.0), (2, 4.0))),
      Vectors.dense(0.0, 3.0, 5.0, 4.0),
      Vectors.dense(0.0, 3.0, 7.0, 4.0)
    ))
```

first element of each vector is zero for dense and empty for sparse. With out this hack significant induces will have first element, because they has different values (zero and empty), but if you convert sparse vector to dense vector significant induces hasn't got first element.

It is clear now?